### PR TITLE
SQL: Fix issue with optimization on queries with ORDER BY/LIMIT

### LINKED ISSUE #40211: SQL: Exception thrown for trivial GROUP BY and LIMIT
```SELECT '1' FROM test GROUP BY 1' LIMIT <x>```
throws an exception:
```Encountered a bug; LocalExec['1'{c}#161=1] is a LocalExec but is not empty```

without the LIMIT it works fine.

### PR DESCRIPTION
**Pull request by matriv**
Previously, when a trival plain `SELECT` or a trivial `SELECT` with
aggregations has also an `ORDER BY` or a `LIMIT` or both, then the
optimization to convert it to a `LocalRelation` was skipped resulting
in exception thrown. E.g.::
```
SELECT 'foo' FROM test LIMIT 10
```
or
```
SELECT 'foo' FROM test GROUP BY 1 ORDER BY 1
```

Fixes: #40211

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 1]
```diff
@@ -1959,6 +1975,19 @@ private static boolean isNotQueryWithFromClauseAndFilterFoldedToFalse(UnaryPlan
             return (!(plan.child() instanceof LocalRelation) || (plan.child() instanceof LocalRelation &&
                 !(((LocalRelation) plan.child()).executable() instanceof EmptyExecutable)));
         }
+
+        private static LocalRelation findLocalRelation(LogicalPlan plan) {
+            Holder<LocalRelation> logicalPlanHolder = new Holder<>();
+            plan.transformDown(l -> {
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: Is `transformDown` really needed here or `forEachDown` is enough?

**Review by astefan — APPROVED** [review]
LGTM. Left one comment.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java (hunk 1 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: Ah, thx, didn't think of that!

**Review by costin — APPROVED** [review]
LGTM.

**Comment by matriv** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Comment by matriv** [comment]
@elasticmachine run elasticsearch-ci/1
@elasticmachine run elasticsearch-ci/2
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/docbldesx
@elasticmachine run elasticsearch-ci/oss-distro-docs
@elasticmachine run elasticsearch-ci/packaging-sample

**Comment by matriv** [comment]
@elasticmachine run elasticsearch-ci/default-distro
@elasticmachine run elasticsearch-ci/bwc

**Comment by matriv** [comment]
Backported to `7.x` with f37f2b5d39676d45706efd29b2fab9650158d5a7
to `7.0` with d118f8529b1e8fda17d16dd21fb4903d6f109956
to `6.7` with 3c008663a2749ff38804700bf0163814369b703f
to `6.6` with 1b30a627d3c1a0bc5fddbe52b80ef7f3c0ce35d1

**Pull request merged**